_posts/2020-09-02-mutiny-redis.adoc

```diff
@@ -0,0 +1,139 @@
+
+Both methods are asynchronous (they return `Uni`), and we need to call the second one for each retrieved key. In other words, we need to iterate over the set of keys, and for each key to invoke an asynchronous action. Finally, we want to collect the result of these asynchronous actions into a `JsonArray`.
+
+Let's start with the beginning; we need the Redis Client instance:
+
+[source, java]
+----
+RedisClient redis = RedisClient.create(vertx, new JsonObject()
+     .put("port", 6379)
+     .put("host", "localhost"));
+----
+
+Note that in Quarkus, you should use the Redis extension directly, which exposes a similar API. Enrico wanted to use the Vert.x Redis Client directly.
+
+Now that we have our client, let's retrieve the list of keys:
+
+[source, java]
+----
+Uni<JsonArray> keys = redis.keys("*")
+----
+
+That produces the `JsonArray`, but we want a stream of keys.  Again, it's an asynchronous method. The returned `Uni` receives the array when it's available. Once received (`onItem`), we can create a stream out of this array:
+
+[source, java]
+----
+Multi<String> keys = redis.keys("*")
+     .onItem().transformToMulti(array -> Multi.createFrom().iterable(array))
+     .onItem().castTo(String.class);
+----
+
+This snippet:
+
+1. retrieves the `JsonArray` containing the keys
+2. creates a `Multi` streaming these keys, it's a `Multi<Object>` as a `JsonArray` is extending `Iterable<Object>`
+3. maps the items from this `Multi` to `String`
+
+At this point, we have a stream of (String) keys. So, we are done with step 1. 
+
+Now, step 2: for each key, we want to retrieve the associated object.
+
+So let's use the `hgetall` method:
+
+[source, java]
+----
+Multi<JsonObject> objects = keys
+  .onItem().transformToUniAndMerge(key -> redis.hgetall(key));
+----
+
+This snippet requires a bit of an explanation. 
+
+For each item of the stream `keys`, we call `hgetall`, which produces a `Uni<JsonObject>`. 
+
+So, we want to transform our key into a Uni (*transformToUni*).
+
+When you have a stream of items and need to invoke an asynchronous action for each item, you must choose how you will merge the results. Mutiny provides two strategies:
+
+- *merge -* as soon as the item produced by the `Uni` is received we send it downstream
+- *concatenate* - we preserve the order of the input stream to be sure that the items are sent downstream in the same order
+
+Let's illustrate this. Imagine we have the keys `1`, `2`, `3` and to the stream `{1, 2, 3}`. Also, let's consider that in our Redis database, the key `1` is associated to `A`, `2` to `B` and `3` to `C`. 
+
+If you use the *merge* strategy, we are retrieving the associated objects we the order is undetermined. We can end up with `{A, C, B}` or `{B, A, C}`. It depends on many factors, such as the latency,  scheduling, load and so on. However, it also means we can retrieve all the associated objects concurrently and produce the resulting stream without taking care of the order.
```

Comment:
"objects we the order" doesn't make sense to me

Reply:
Is it meant to be "where"?

Reply:
Remove the extra space after "latency,"

Reply:
It meant to be: "If you use the *merge* strategy, we are retrieving the associated objects in an undetermined order."